### PR TITLE
Forbid unused locals/parameters in compiler

### PR DESCRIPTION
**Pull request by andy-ms**
Mostly I just removed any parameter that wasn't used, and then removed it from callers. If a function is used as a callback, and other callbacks use that parameter, I use an `_underscored` parameter name.

TODO: The same thing for harness/server/services, and probably use tsconfig inheritance.

### CODE REVIEW

File: src/compiler/comments.ts [hunk 1]
```diff
@@ -182,13 +182,13 @@ namespace ts {
             }
         }
 
-        function emitTripleSlashLeadingComment(commentPos: number, commentEnd: number, kind: SyntaxKind, hasTrailingNewLine: boolean, rangePos: number) {
+        function emitTripleSlashLeadingComment(commentPos: number, commentEnd: number, _kind: SyntaxKind, hasTrailingNewLine: boolean, rangePos: number) {
```
Comment (mhegazy) [inline src/compiler/comments.ts]: why `_kind`? seems it is used two lines below.

File: src/compiler/core.ts [hunk 2]
```diff
@@ -903,7 +903,7 @@ namespace ts {
             return t => compose(a(t));
         }
         else {
-            return t => u => u;
+            return _t => u => u;
```
Comment (mhegazy) [inline src/compiler/core.ts]: consider making it jsut `_`

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ namespace ts {
     }
 
     const id = (s: SourceFile) => s;
-    const nullTransformers: Transformer[] = [ctx => id];
+    const nullTransformers: Transformer[] = [_ctx => id];
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: `_` instead of `_ctx`

File: src/compiler/moduleNameResolver.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ namespace ts {
 
     /* @internal */
     export function trace(host: ModuleResolutionHost, message: DiagnosticMessage, ...args: any[]): void;
-    export function trace(host: ModuleResolutionHost, message: DiagnosticMessage): void {
+    export function trace(host: ModuleResolutionHost, _message: DiagnosticMessage): void {
```
Comment (mhegazy) [inline src/compiler/moduleNameResolver.ts]: consider removing `_message` all together from the implementation signature,  `export function trace(host: ModuleResolutionHost): void { .. }`

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -575,7 +575,7 @@ namespace ts {
         export function parseSourceFile(fileName: string, _sourceText: string, languageVersion: ScriptTarget, _syntaxCursor: IncrementalParser.SyntaxCursor, setParentNodes?: boolean, scriptKind?: ScriptKind): SourceFile {
```
Comment (mhegazy) [inline src/compiler/parser.ts]: why do these start with `_`?

File: src/compiler/sys.ts [hunk 6]
```diff
@@ -595,9 +591,9 @@ namespace ts {
                 getExecutingFilePath: () => ChakraHost.executingFile,
                 getCurrentDirectory: () => ChakraHost.currentDirectory,
                 getDirectories: ChakraHost.getDirectories,
-                getEnvironmentVariable: ChakraHost.getEnvironmentVariable || ((name: string) => ""),
+                getEnvironmentVariable: ChakraHost.getEnvironmentVariable || ((_name: string) => ""),
```
Comment (mhegazy) [inline src/compiler/sys.ts]: do not think you need ht type annotation here.

**Comment by andy-ms** [comment]
`typescript.d.ts` LKG is identical whether built on master or in this branch. Assuming you didn't mean for me to check in a new LKG.

**Pull request merged**